broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/NewPartitionTransitionImpl.java

```diff
@@ -63,75 +54,99 @@ public void updateTransitionContext(final PartitionTransitionContext transitionC
     return transitionTo(INACTIVE_TERM, Role.INACTIVE);
   }
 
+  @Override
+  public void setConcurrencyControl(final ConcurrencyControl concurrencyControl) {
+    this.concurrencyControl = requireNonNull(concurrencyControl);
+  }
+
+  @Override
+  public void updateTransitionContext(final PartitionTransitionContext transitionContext) {
+    context = transitionContext;
+  }
+
   public ActorFuture<Void> transitionTo(final long term, final Role role) {
-    LOG.info(format("Transition to %s on term %d requested.", role, term));
+    LOG.info("Transition to {} on term {} requested.", role, term);
 
     // notify steps immediately that a transition is coming; steps are encouraged to cancel any
     // ongoing activity at this point in time
     steps.forEach(step -> step.onNewRaftRole(context, role));
 
     final ActorFuture<Void> nextTransitionFuture = concurrencyControl.createFuture();
-
-    concurrencyControl.run(
-        () -> {
-          if (currentTransition != null) {
-            LOG.info(
-                format(
-                    "Transition to %s on term %d requested while another transition is still running",
-                    role, term));
-            currentTransition.cancel(); // this will drop any subsequent transition steps
-
-            // schedule new transition as soon as the current step of the current transition
-            // has completed
-            concurrencyControl.runOnCompletion(
-                currentTransitionFuture,
-                (ok, error) -> cleanupLastTransition(nextTransitionFuture, term, role));
-
-          } else {
-            cleanupLastTransition(nextTransitionFuture, term, role);
+    final var nextTransition =
+        new PartitionTransitionProcess(steps, concurrencyControl, context, term, role);
+    nextTransitionFuture.onComplete(
+        (v, error) -> {
```

Comment:
I would also move this inside `concurrencyControl.run()`, as this should be executed on the transitions's actor not the caller's actor.

Reply:
Thanks you're right